File: Javascript/90. Subsets II.js
/**
 * Link: https://leetcode.com/problems/subsets-ii/
 * Author: Dung Nguyen Tien (Chris)
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsetsWithDup = function(nums) {
    nums.sort(function(a, b) {return a - b;});
    
    var res = [];
    backTrack(res, [], nums, 0);
    
    return res;
};

function backTrack(res, tmp, nums, start) {
    res.push(tmp.slice(0));
    
    for (var i = start; i < nums.length; i++) {
        // skip the duplicated item
        if (i > start && nums[i] == nums[i - 1]) continue;
        
        tmp.push(nums[i]);
        backTrack(res, tmp, nums, i + 1);
        tmp.pop();
    }
}

File: C#/90. Subsets II.cs
public class Solution {
    public IList<IList<int>> SubsetsWithDup(int[] nums) {
        Array.Sort(nums);
        
        var res = new List<IList<int>>();
        
        Backtrack(res, new List<int>(), nums, 0);
        
        return res;
    }
    
    public void Backtrack(List<IList<int>> res, List<int> cur, int[] nums, int index) {
        res.Add(new List<int>(cur));
        
        for (var i = index; i < nums.Length; i++) {
            if (i > index && nums[i] == nums[i - 1]) continue;
            
            cur.Add(nums[i]);
            Backtrack(res, cur, nums, i + 1);
            cur.RemoveAt(cur.Count - 1);
        }
    }
}